kroxylicious-operator/src/main/java/io/kroxylicious/kubernetes/operator/ClusterService.java

```diff
@@ -49,82 +46,20 @@ static String serviceName(VirtualKafkaCluster cluster) {
-            if (clusters.get(clusterNum).getMetadata().getName().equals(cluster.getMetadata().getName())) {
-                if (SharedKafkaProxyContext.isBroken(context, cluster)) {
-                    return Map.of();
-                }
-                int startPort = 9292 + (100 * clusterNum);
-                int numBrokerPorts = 4;
-                return IntStream.range(startPort, startPort + numBrokerPorts).boxed()
-                        .collect(Collectors.<Integer, Integer, String, TreeMap<Integer, String>> toMap(
-                                portNum -> portNum,
-                                portNum -> cluster.getMetadata().getName() + "-" + portNum,
-                                (v1, v2) -> {
-                                    throw new IllegalStateException();
-                                },
-                                TreeMap::new));
-            }
-        }
-        throw new IllegalArgumentException("Couldn't find cluster with name " + cluster.getMetadata().getName());
-    }
-
-    protected Service clusterService(KafkaProxy primary,
-                                     Context<KafkaProxy> context,
-                                     VirtualKafkaCluster cluster) {
-        // @formatter:off
-        var serviceSpecBuilder = new ServiceBuilder()
-                .withNewMetadata()
-                    .withName(serviceName(cluster))
-                    .withNamespace(primary.getMetadata().getNamespace())
-                    .addToLabels(standardLabels(primary))
-                    .addNewOwnerReferenceLike(ResourcesUtil.ownerReferenceTo(primary)).endOwnerReference()
-                .endMetadata()
-                .withNewSpec()
-                    .withSelector(ProxyDeployment.podLabels(primary));
-        for (var portNumEntry : clusterPorts( context, cluster).entrySet()) {
-            serviceSpecBuilder = serviceSpecBuilder
-                    .addNewPort()
-                        .withName(portNumEntry.getValue())
-                        .withPort(portNumEntry.getKey())
-                        .withTargetPort(new IntOrString(portNumEntry.getKey()))
-                        .withProtocol("TCP")
-                    .endPort();
-        }
-
-        return serviceSpecBuilder
-                .endSpec()
-                .build();
-        // @formatter:on
-    }
-
     @Override
     public Map<String, Service> desiredResources(
                                                  KafkaProxy primary,
                                                  Context<KafkaProxy> context) {
-        return ResourcesUtil.clustersInNameOrder(context)
+        List<VirtualKafkaCluster> clusters = ResourcesUtil.clustersInNameOrder(context).toList();
+        Set<KafkaProxyIngress> ingresses = context.getSecondaryResources(KafkaProxyIngress.class);
+        ProxyIngressLayout layout = ProxyIngressLayout.layout(primary, clusters, ingresses);
```

Comment:
Note: we do the secondary lookup resources here so that the ClusterService and ProxyIngressLayout has a consistent view of what secondary resources exist. The integration test showed that the secondary resources can change during a single reconciliation, so if ProxyIngressLayout looked up the resources itself we can end up with inconsistencies.

Reply:
this also interfered with trying to cache the layout calculation, I think it's better to just have each sub-resource recalculate it and they should eventually converge